### DISCUSSION TITLE
FIR filtering on ESP32

### DISCUSSION THREAD

**Discussion by Tarun-Sai-Chowdary** [pschatzmann/arduino-audio-tools#394] [Q&A] [answered]
Hey I was working on INMP441 with esp32 and want to get frequency range from 20 to 200 hz only. I used streams i2s-filter-i2s example code but haven't got the filtered output. As the sampling rate is varied inmp441 has inbuilt band pass filter that filters some of the frequency. Can some one help me figure it out? And also how to select sampling frequency to get the filter coefficients in matlab or tfilter.

**Comment by pschatzmann** [comment, marked as answer]
The [documentation in the Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Filters) contains a link to a filter design tool.
I don't think that you can use the built in low pass filter of the INMP441: You would need to select a sampling rate of 300-400 hz, but this will not give any good audio!

**Reply by Tarun-Sai-Chowdary** [reply]
That is the issue, when I tried i2s-i2s-2 example keeping the sampling rate to 400 it was acting as an all pass filter and audio was really good, when changed to 8000 sampling rate its acting like an low pass filter with 2000hz cutoff frequency. I tried all this without fir filter. If used static kinda noise is being added. I will add the code which I used.

**Reply by Tarun-Sai-Chowdary** [reply]
#include "AudioTools.h"
 
uint16_t sample_rate=4000; // Try with different Sampling_rates as per requirement
// for 100 sample_rate there is no output at all
// for 200 sample_rate the output is showing cutoff_frequency at 3300hz
// for 400 sample_rate the output is showing 1st cutoff at 2000hz and next cutoff at 6000hz(wierd_behaviour)
// for 500 sample_rate no output
// for 800 sample_rate the output is showing cutoff frequency at 13000hz
// for 1000 sample_rate thee is no output
// for 2000 sample_rate 16bit the output is showing cutoff frequency at 3000hz
// for 8000 sample_rate 16bit the output is showing cutoff frequency at 2000hz
// for 16000 sample_rate the output is showing cutoff frequency at 4000hz
// for 32000 sample_rate the output is showing cutoff frequency at 8000hz
// for 44100 sample_rate the output is showing cutoff frequency at 13000hz

// for 2000 sample_rate 32bit the output is showing cutoff frequency at 5000hz
// for 8000 sample_rate 32bit the output is showing cutoff frequency at 4000hz
// for 16000 sample_rate 32bit the output is showing cutoff frequency at 7500hz


// for 400 sample_rate 24 bi the output is showing cutoff frequncy at 10000hz
// for 8000 smaple_rate 24bit the output is showing cutoff frequnecy at 3500hz
// for 16000 sample_rate 24bit the output is showing cutoff frequnecy at 7000hz



uint16_t channels = 1;
uint16_t bits_per_sample = 16; // or try with 24 or 32
int I2S_READ_LEN = 512;

I2SStream I2S_in;
I2SStream I2S_out; 
uint16_t * i2s_read_buff = (uint16_t *)calloc(I2S_READ_LEN, sizeof(uint16_t));
//StreamCopy copier(I2S_out, I2S_in); // copies sound into I2S_in to I2S_out


void setup(void) {  
  
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Info); 

  Serial.println("starting I2S...");
  auto config_in = I2S_in.defaultConfig(RX_MODE);
  config_in.sample_rate = sample_rate; 
  config_in.bits_per_sample = bits_per_sample; 
  config_in.i2s_format = I2S_STD_FORMAT;
  //config_in.use_apll=1;
  // config_in.pin_mck=0;
  config_in.port_no = 0;
  config_in.pin_bck = 12;
  config_in.pin_ws = 13;
  config_in.pin_data = 15;  
  I2S_in.begin(config_in);

  auto config_out = I2S_out.defaultConfig(TX_MODE);
  config_out.sample_rate = sample_rate; 
  config_out.bits_per_sample = bits_per_sample;
  config_out.i2s_format = I2S_STD_FORMAT;
  //config_out.use_apll=1;
  config_out.port_no = 1;
  config_out.pin_bck = 14;
  config_out.pin_ws = 5;
  config_out.pin_data = 22;
  I2S_out.begin(config_out);

  Serial.println("I2S started...");
}
 
void loop() {
  I2S_in.readBytes((uint8_t*)i2s_read_buff, I2S_READ_LEN);
  //for(int i=0; i<I2S_READ_LEN; i++){
//    i2s_read_buff[i] *= 100;
 // }
  I2S_out.write((uint8_t*)i2s_read_buff, I2S_READ_LEN);
}

**Comment by pschatzmann** [comment]
Did you measure the effective sample rate: I would be surprised if anything below 8000 would work correctly

**Comment by Tarun-Sai-Chowdary** [comment]
What is this effective sampling rate? I have no idea about it.

**Comment by pschatzmann** [comment]
Just add a counter to the loop that is resetting every second and count the number of frames that you process (I2S_READ_LEN/(sizeof(int16_t)*2))

**Reply by Tarun-Sai-Chowdary** [reply]
Sorry I didn't understand the counter and am not aware of that. Can you please elaborate it?

**Comment by pschatzmann** [comment]
The sample rate is the number of samples per second: So you can measure this in you loop.
I2S has always 2 channels and 16 bits = 2 bytes: so you need to divide the processed bytes by 4.

**Reply by Tarun-Sai-Chowdary** [reply]
Correct me if  I am wrong, When we are using I2S first the data is loaded to the dma buffer and when it interrupts the cpu the data is transferred to the i2s_read_buff, the thing which I didn't understand is by reading the buffer as it is dynamically changing can we know how frames of data we captured in one sec.

**Comment by pschatzmann** [comment]
I don't think anything is dynamically changing: You just empty the buffer at full speed and this is limited by the speed the buffer is filled: This should be pretty consistent...

**Reply by Tarun-Sai-Chowdary** [reply]
void loop() {
  unsigned long start = millis();
  I2S_in.readBytes((uint8_t*)i2s_read_buff, I2S_READ_LEN);
  Serial.println(millis()-start);
}

I tried, using this loop to get the time taken to fill the buffer of 512 size (I2S_READ_LEN = 512) , with 400 sampling rate it came 250ms and with 8000 sampling rate it came 64ms. With the 400 sampling rate it doesn't make sense for four cycles( implies 1 sec time) it was collecting 512 samples instead of 400 samples. I am unable to figure out the code for that clearing buffer  and calculating the samples, if possible can you attach that code here.

**Comment by pschatzmann** [comment]
That was my point: you can't go that low!
You don't need to clear any buffer: 512 bytes are 128 frames (=samples)

**Reply by Tarun-Sai-Chowdary** [reply]
But the audio was very clear even at that sampling rate which was mystery, even though the effective sampling rate is more than that.

**Comment by pschatzmann** [comment]
I guess it is just the other way round: it was clear because the effective sampling rate was much higher then you thought it is...
